File: android/app/src/main/java/com/jianxing/browser/data/HistoryStore.kt
package com.jianxing.browser.data

import android.content.Context
import android.content.SharedPreferences
import com.jianxing.browser.model.HistoryEntry
import org.json.JSONArray
import org.json.JSONObject
import java.net.URI
import java.util.UUID

class HistoryStore(context: Context) {
    private val prefs: SharedPreferences =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    private var cached: Snapshot? = null

    @Synchronized
    fun list(query: String = ""): List<HistoryEntry> {
        val q = query.trim().lowercase()
        return load().entries
            .filter { e ->
                q.isEmpty() ||
                    e.title.lowercase().contains(q) ||
                    e.url.lowercase().contains(q) ||
                    e.host.lowercase().contains(q)
            }
            .sortedByDescending { it.visitedAt }
    }

    @Synchronized
    fun count(): Int = load().entries.size

    @Synchronized
    fun getRevision(): Int = load().revision

    @Synchronized
    fun setRevision(revision: Int) {
        val snap = load()
        save(snap.copy(revision = maxOf(0, revision)))
    }

    @Synchronized
    fun exportForSync(): Snapshot = load()

    @Synchronized
    fun mergeFromSync(
        remoteEntries: List<HistoryEntry>,
        remoteDeletedIds: List<String>,
        remoteClearedAt: Long,
        remoteRevision: Int
    ): Boolean {
        val local = load()
        val deleted = pruneDeletedIds(local.deletedIds + remoteDeletedIds)
        val clearedAt = maxOf(local.clearedAt, remoteClearedAt.coerceAtLeast(0L))
        val byId = LinkedHashMap<String, HistoryEntry>()
        for (entry in local.entries + remoteEntries) {
            val normalized = normalize(entry) ?: continue
            if (deleted.contains(normalized.id)) continue
            if (clearedAt > 0 && normalized.visitedAt <= clearedAt) continue
            val prev = byId[normalized.id]
            byId[normalized.id] = if (prev == null) normalized else pickNewer(prev, normalized)
        }
        val merged = byId.values.sortedByDescending { it.visitedAt }.take(MAX_ENTRIES)
        val revision = maxOf(local.revision, remoteRevision.coerceAtLeast(0))
        val next = Snapshot(
            entries = merged,
            deletedIds = deleted,
            clearedAt = clearedAt,
            revision = revision
        )
        val changed = !payloadEqual(local, next)
        save(next)
        return changed
    }

    @Synchronized
    fun record(url: String, title: String?): HistoryEntry? {
        val href = url.trim()
        if (!href.startsWith("http://") && !href.startsWith("https://")) return null
        val now = System.currentTimeMillis()
        val host = try {
            URI(href).host?.lowercase().orEmpty()
        } catch (_: Exception) {
            ""
        }
        val pageTitle = title?.trim().orEmpty().ifBlank { host.ifBlank { href } }
        val snap = load()
        val items = snap.entries.toMutableList()
        val lastSame = items.firstOrNull { it.url == href }
        if (lastSame != null && now - lastSame.visitedAt < DEDUPE_MS) {
            val updated = lastSame.copy(
                title = if (pageTitle.isNotBlank() && pageTitle != href) pageTitle else lastSame.title,
                visitedAt = now
            )
            items.removeAll { it.id == lastSame.id }
            items.add(0, updated)
            save(snap.copy(entries = items.take(MAX_ENTRIES)))
            return updated
        }
        val entry = HistoryEntry(
            id = "h_${UUID.randomUUID().toString().replace("-", "").take(12)}",
            url = href,
            title = pageTitle,
            host = host,
            visitedAt = now
        )
        items.add(0, entry)
        save(snap.copy(entries = items.take(MAX_ENTRIES)))
        return entry
    }

    @Synchronized
    fun remove(id: String): Boolean {
        val snap = load()
        val next = snap.entries.filter { it.id != id }
        if (next.size == snap.entries.size) return false
        save(
            snap.copy(
                entries = next,
                deletedIds = pruneDeletedIds(snap.deletedIds + id)
            )
        )
        return true
    }

    @Synchronized
    fun clear() {
        val snap = load()
        save(
            snap.copy(
                entries = emptyList(),
                deletedIds = pruneDeletedIds(snap.deletedIds + snap.entries.map { it.id }),
                clearedAt = System.currentTimeMillis()
            )
        )
    }

    data class Snapshot(
        val entries: List<HistoryEntry>,
        val deletedIds: List<String> = emptyList(),
        val clearedAt: Long = 0L,
        val revision: Int = 0
    )

    private fun load(): Snapshot {
        cached?.let { return it }
        val json = prefs.getString(KEY_ENTRIES, null) ?: return Snapshot(emptyList()).also { cached = it }
        return try {
            val root = JSONObject(json)
            val arr = root.optJSONArray("entries") ?: JSONArray()
            val entries = (0 until arr.length()).mapNotNull { i ->
                normalize(parseEntry(arr.getJSONObject(i)))
            }
            val deletedArr = root.optJSONArray("deletedIds") ?: JSONArray()
            val deleted = (0 until deletedArr.length()).map { deletedArr.optString(it) }
            Snapshot(
                entries = entries,
                deletedIds = pruneDeletedIds(deleted),
                clearedAt = root.optLong("clearedAt", 0L).coerceAtLeast(0L),
                revision = root.optInt("revision", 0).coerceAtLeast(0)
            ).also { cached = it }
        } catch (_: Exception) {
            Snapshot(emptyList()).also { cached = it }
        }
    }

    private fun save(snap: Snapshot) {
        cached = snap
        val arr = JSONArray()
        snap.entries.forEach { arr.put(entryToJson(it)) }
        val deleted = JSONArray()
        snap.deletedIds.forEach { deleted.put(it) }
        prefs.edit()
            .putString(
                KEY_ENTRIES,
                JSONObject()
                    .put("version", 2)
                    .put("revision", snap.revision)
                    .put("clearedAt", snap.clearedAt)
                    .put("deletedIds", deleted)
                    .put("entries", arr)
                    .toString()
            )
            .apply()
    }

    companion object {
        private const val PREFS_NAME = "jianxing_history"
        private const val KEY_ENTRIES = "entries_json"
        private const val MAX_ENTRIES = 2000
        private const val MAX_DELETED_IDS = 3000
        private const val DEDUPE_MS = 2000L

        fun parseEntry(o: JSONObject): HistoryEntry =
            HistoryEntry(
                id = o.optString("id"),
                url = o.optString("url"),
                title = o.optString("title"),
                host = o.optString("host"),
                visitedAt = o.optLong("visitedAt", 0L)
            )

        fun parseEntries(arr: JSONArray): List<HistoryEntry> =
            (0 until arr.length()).mapNotNull { i ->
                try {
                    normalize(parseEntry(arr.getJSONObject(i)))
                } catch (_: Exception) {
                    null
                }
            }

        fun parseDeletedIds(arr: JSONArray?): List<String> {
            if (arr == null) return emptyList()
            return (0 until arr.length()).map { arr.optString(it) }.filter { it.isNotBlank() }
        }

        fun entryToJson(e: HistoryEntry): JSONObject =
            JSONObject()
                .put("id", e.id)
                .put("url", e.url)
                .put("title", e.title)
                .put("host", e.host)
                .put("visitedAt", e.visitedAt)

        fun payloadEqual(a: Snapshot, b: Snapshot): Boolean {
            if (a.clearedAt != b.clearedAt) return false
            if (a.deletedIds.toSet() != b.deletedIds.toSet()) return false
            if (a.entries.size != b.entries.size) return false
            val left = a.entries.sortedBy { it.id }
            val right = b.entries.sortedBy { it.id }
            return left == right
        }

        private fun normalize(raw: HistoryEntry): HistoryEntry? {
            val id = raw.id.trim()
            val url = raw.url.trim()
            if (id.isEmpty()) return null
            if (!url.startsWith("http://") && !url.startsWith("https://")) return null
            val host = raw.host.trim().lowercase().ifBlank {
                try {
                    URI(url).host?.lowercase().orEmpty()
                } catch (_: Exception) {
                    ""
                }
            }
            val title = raw.title.trim().ifBlank { host.ifBlank { url } }
            return raw.copy(id = id, url = url, title = title, host = host)
        }

        private fun pickNewer(a: HistoryEntry, b: HistoryEntry): HistoryEntry {
            if (b.visitedAt != a.visitedAt) {
                return if (b.visitedAt > a.visitedAt) b else a
            }
            val aWeak = a.title.isBlank() || a.title == a.url || a.title == a.host
            val bWeak = b.title.isBlank() || b.title == b.url || b.title == b.host
            if (aWeak && !bWeak) return b
            if (bWeak && !aWeak) return a
            return if (b.title.length >= a.title.length) b else a
        }

        private fun pruneDeletedIds(ids: List<String>): List<String> {
            val unique = ArrayList<String>()
            val seen = HashSet<String>()
            for (i in ids.indices.reversed()) {
                val id = ids[i].trim()
                if (id.isEmpty() || !seen.add(id)) continue
                unique.add(id)
                if (unique.size >= MAX_DELETED_IDS) break
            }
            unique.reverse()
            return unique
        }
    }
}

File: src/main/history-store.ts
import { app } from 'electron';
import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'fs';
import { dirname, join } from 'path';
import { randomBytes } from 'crypto';

export interface HistoryEntry {
  id: string;
  url: string;
  title: string;
  host: string;
  visitedAt: number;
}

export type HistorySyncPayload = {
  entries: HistoryEntry[];
  deletedIds: string[];
  clearedAt: number;
};

type StoreFile = {
  version: 1 | 2;
  revision?: number;
  clearedAt?: number;
  deletedIds?: string[];
  entries: HistoryEntry[];
};

const MAX_ENTRIES = 2000;
const MAX_DELETED_IDS = 3000;
const DEDUPE_MS = 2000;

function storePath(): string {
  return join(app.getPath('userData'), 'history.json');
}

function newId(): string {
  return `h_${Date.now().toString(36)}_${randomBytes(3).toString('hex')}`;
}

function parseHost(url: string): string {
  try {
    return new URL(url).hostname.toLowerCase();
  } catch {
    return '';
  }
}

function normalizeEntry(raw: Partial<HistoryEntry> | null | undefined): HistoryEntry | null {
  if (!raw || typeof raw !== 'object') return null;
  const id = String(raw.id || '').trim();
  const url = String(raw.url || '').trim();
  if (!id || !/^https?:\/\//i.test(url)) return null;
  const host = String(raw.host || '').trim().toLowerCase() || parseHost(url);
  const title = String(raw.title || '').trim() || host || url;
  const visitedAt = Number(raw.visitedAt) || 0;
  return { id, url, title, host, visitedAt };
}

function pruneDeletedIds(ids: string[]): string[] {
  const unique: string[] = [];
  const seen = new Set<string>();
  for (let i = ids.length - 1; i >= 0; i -= 1) {
    const id = String(ids[i] || '').trim();
    if (!id || seen.has(id)) continue;
    seen.add(id);
    unique.push(id);
    if (unique.length >= MAX_DELETED_IDS) break;
  }
  unique.reverse();
  return unique;
}

export function historyPayloadEqual(
  a: HistorySyncPayload,
  b: HistorySyncPayload
): boolean {
  return (
    JSON.stringify(normalizeHistoryPayload(a)) ===
    JSON.stringify(normalizeHistoryPayload(b))
  );
}

function normalizeHistoryPayload(payload: HistorySyncPayload): unknown {
  return {
    clearedAt: Number(payload.clearedAt) || 0,
    deletedIds: [...new Set(payload.deletedIds || [])].filter(Boolean).sort(),
    entries: [...(payload.entries || [])]
      .map((e) => ({
        id: e.id,
        url: e.url,
        title: e.title,
        host: e.host,
        visitedAt: e.visitedAt,
      }))
      .sort((x, y) => x.id.localeCompare(y.id) || x.visitedAt - y.visitedAt),
  };
}

function pickNewer(a: HistoryEntry, b: HistoryEntry): HistoryEntry {
  if (b.visitedAt !== a.visitedAt) {
    return b.visitedAt > a.visitedAt ? b : a;
  }
  const aWeak = !a.title || a.title === a.url || a.title === a.host;
  const bWeak = !b.title || b.title === b.url || b.title === b.host;
  if (aWeak && !bWeak) return b;
  if (bWeak && !aWeak) return a;
  return b.title.length >= a.title.length ? b : a;
}

export class HistoryStore {
  private entries: HistoryEntry[] = [];
  private deletedIds: string[] = [];
  private clearedAt = 0;
  private revision = 0;

  constructor() {
    this.load();
  }

  private load(): void {
    const path = storePath();
    if (!existsSync(path)) {
      this.entries = [];
      this.deletedIds = [];
      this.clearedAt = 0;
      this.revision = 0;
      return;
    }
    try {
      const raw = JSON.parse(readFileSync(path, 'utf8')) as StoreFile;
      this.entries = Array.isArray(raw.entries)
        ? raw.entries.map((e) => normalizeEntry(e)).filter((e): e is HistoryEntry => Boolean(e))
        : [];
      this.deletedIds = pruneDeletedIds(Array.isArray(raw.deletedIds) ? raw.deletedIds : []);
      this.clearedAt = Math.max(0, Number(raw.clearedAt) || 0);
      this.revision = Math.max(0, Number(raw.revision) || 0);
    } catch {
      this.entries = [];
      this.deletedIds = [];
      this.clearedAt = 0;
      this.revision = 0;
    }
  }

  private persist(): void {
    const path = storePath();
    mkdirSync(dirname(path), { recursive: true });
    const data: StoreFile = {
      version: 2,
      revision: this.revision,
      clearedAt: this.clearedAt,
      deletedIds: this.deletedIds,
      entries: this.entries,
    };
    writeFileSync(path, JSON.stringify(data), 'utf8');
  }

  list(query?: string): HistoryEntry[] {
    const q = String(query || '')
      .trim()
      .toLowerCase();
    let rows = this.entries.map((e) => ({ ...e }));
    if (q) {
      rows = rows.filter(
        (e) =>
          e.title.toLowerCase().includes(q) ||
          e.url.toLowerCase().includes(q) ||
          e.host.toLowerCase().includes(q)
      );
    }
    return rows.sort((a, b) => b.visitedAt - a.visitedAt);
  }

  count(): number {
    return this.entries.length;
  }

  get(id: string): HistoryEntry | undefined {
    const found = this.entries.find((e) => e.id === id);
    return found ? { ...found } : undefined;
  }

  getRevision(): number {
    return this.revision;
  }

  setRevision(revision: number): void {
    this.revision = Math.max(0, Number(revision) || 0);
    this.persist();
  }

  exportForSync(): HistorySyncPayload {
    return {
      entries: this.entries.map((e) => ({ ...e })),
      deletedIds: [...this.deletedIds],
      clearedAt: this.clearedAt,
    };
  }

  mergeFromSync(
    remoteEntries: Partial<HistoryEntry>[],
    remoteDeletedIds: string[] = [],
    remoteClearedAt = 0,
    remoteRevision = 0
  ): { changed: boolean } {
    const before = JSON.stringify(normalizeHistoryPayload(this.exportForSync()));
    const deleted = pruneDeletedIds([...this.deletedIds, ...remoteDeletedIds]);
    const clearedAt = Math.max(this.clearedAt, Number(remoteClearedAt) || 0);
    const byId = new Map<string, HistoryEntry>();
    for (const raw of [...this.entries, ...remoteEntries]) {
      const entry = normalizeEntry(raw);
      if (!entry) continue;
      if (deleted.includes(entry.id)) continue;
      if (clearedAt > 0 && entry.visitedAt <= clearedAt) continue;
      const prev = byId.get(entry.id);
      byId.set(entry.id, prev ? pickNewer(prev, entry) : entry);
    }
    const merged = [...byId.values()]
      .sort((a, b) => b.visitedAt - a.visitedAt)
      .slice(0, MAX_ENTRIES);
    this.entries = merged;
    this.deletedIds = deleted;
    this.clearedAt = clearedAt;
    if (typeof remoteRevision === 'number' && remoteRevision > this.revision) {
      this.revision = remoteRevision;
    }
    const after = JSON.stringify(normalizeHistoryPayload(this.exportForSync()));
    this.persist();
    return { changed: before !== after };
  }

  record(url: string, title?: string): HistoryEntry | null {
    const href = String(url || '').trim();
    if (!/^https?:\/\//i.test(href)) return null;

    const now = Date.now();
    const host = parseHost(href);
    const pageTitle = String(title || '').trim() || host || href;
    const lastSame = this.entries.find((e) => e.url === href);
    if (lastSame && now - lastSame.visitedAt < DEDUPE_MS) {
      if (pageTitle && pageTitle !== lastSame.title && pageTitle !== href) {
        lastSame.title = pageTitle;
      }
      lastSame.visitedAt = now;
      this.entries = [
        lastSame,
        ...this.entries.filter((e) => e.id !== lastSame.id),
      ];
      this.persist();
      return { ...lastSame };
    }

    const entry: HistoryEntry = {
      id: newId(),
      url: href,
      title: pageTitle,
      host,
      visitedAt: now,
    };
    this.entries.unshift(entry);
    if (this.entries.length > MAX_ENTRIES) {
      this.entries = this.entries.slice(0, MAX_ENTRIES);
    }
    this.persist();
    return { ...entry };
  }

  updateLatestTitle(url: string, title: string): void {
    const href = String(url || '').trim();
    const pageTitle = String(title || '').trim();
    if (!href || !pageTitle) return;
    const latest = this.entries.find((e) => e.url === href);
    if (!latest || latest.title === pageTitle) return;
    latest.title = pageTitle;
    this.persist();
  }

  remove(id: string): { ok: boolean; error?: string } {
    const idx = this.entries.findIndex((e) => e.id === id);
    if (idx < 0) return { ok: false, error: '记录不存在' };
    this.entries.splice(idx, 1);
    this.deletedIds = pruneDeletedIds([...this.deletedIds, id]);
    this.persist();
    return { ok: true };
  }

  clear(): { ok: boolean } {
    this.deletedIds = pruneDeletedIds([
      ...this.deletedIds,
      ...this.entries.map((e) => e.id),
    ]);
    this.clearedAt = Date.now();
    this.entries = [];
    this.persist();
    return { ok: true };
  }
}
